src/processor/delegate.rs

```diff
@@ -63,11 +63,21 @@ pub fn process_delegate(
     load_owned_pda(delegated_account, &crate::id(), "delegated account")?;
     load_program(system_program, system_program::id(), "system program")?;
 
+    msg!("Delegating: {}", delegated_account.key);
+
     // Validate seeds if the delegate account is not on curve, i.e. is a PDA
+    // If the owner is the system program, we check if the account is derived from the delegation program,
+    // allowing delegation of escrow accounts
     if !is_on_curve(delegated_account.key) {
         let seeds_to_validate: Vec<&[u8]> = args.seeds.iter().map(|v| v.as_slice()).collect();
+        let program_id = if owner_program.key.eq(&system_program::id()) {
```

Comment:
I think this works, just looks a bit hacky though, hope there's no security issues, I couldn't think of hack from top of my head. 

What I just thought we could do is separate:
owner_program and undelegation_target. Right now owner_program == undelegation_target, maybe we could give user power to where he undelegates after, like to system_program. But this is also has unpretty edge cases, like undelegation_target=system_program, means account has to be zeroed out

Reply:
Agreed, that was the other option I was considering, as it would allow delegating system program PDAs from other programs. I ended up not pursuing it because it would require changing the delegation metadata to store the undelegation_record. There's also an edge case where undelegation isn't possible; e.g, if a PDA owned by the system program changes ownership and is resized to contain data, it can't be undelegated back to the system program. Restricting delegation of system program PDAs to the delegation program keeps things controlled, since we don’t have an instruction to resize or add data to escrow accounts. We can revisit this in the future.